Refactor podvm workflows and enable on libvirt e2e

wainersm opened the pull request:
The [libvirt e2e workflow on pull request](https://github.com/confidential-containers/cloud-api-adaptor/actions/workflows/e2e_on_pull.yaml) currently does not create a podvm, instead it uses the pre-built images. We have two problems here: 1) podvm built images are usually old and outdated; 2) not building on pull request means changes on podvm aren't tested on CI.

By the time I introduced the libvirt e2e workflow in PR #1318, @stevenhorsman mentioned on Slack that it would be better to build the podvm too. I agreed, and despite its build takes a lot of time to complete, then I am sending these changes to exactly enable podvm build alongside the libvirt e2e workflow.

Currently the [podvm_builder](https://github.com/confidential-containers/cloud-api-adaptor/tree/main/.github/workflows/podvm_builder.yaml), [podvm_binaries](https://github.com/confidential-containers/cloud-api-adaptor/tree/main/.github/workflows/podvm_binaries.yaml) and [podvm](https://github.com/confidential-containers/cloud-api-adaptor/tree/main/.github/workflows/podvm_builder.yaml) workflows are triggered and gated on release time to publish images on quay.io. I wanted to reuse those workflows on libvirt e2e, so the first commits are meant to prepare the workflows for reuse.

The podvm images created on the libvirt e2e workflow will be published on ghcr.io with the `ci-prN` (where `N` is the PR number) tag, so not mixing up with the official images on quay.io. BTW, the caa images for that workflow are already published this way. I've a script to clean up the ghcr.io from images for PRs closed, I should contribute it soon...

OK, I tested those changes on my cloud-api-adaptor fork, on https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/6173089285 . Unfortunately as libvirt e2e runs on `pull_request_target` we will only be able to fully test it after merging. The podvm release workflow couldn't be tested, unfortunately, and I'm okay if we find prudent to not merge this changes before the 0.8.0 release.

Caching... it has been discussed at least in https://github.com/confidential-containers/cloud-api-adaptor/issues/1410 on how to use cache to speed up the binaries build by leverage caching or another similar mechanism. At begin I thought in implement something on that direction but then I realized it would be too much changes (well, I don't have a solid proposal too) so I preferred to stick with the silly approach took here: it will simple build the builder, binaries and finally podvms all the time. And this whole process can take ~1h.... that to say it is in my radar how to improve that situation on follow up changes.

stevenhorsman commented on .github/workflows/podvm.yaml:
```diff
@@ -55,12 +55,25 @@ jobs:
+
+        case "$event_name" in
+        release)
+          tags="${img}:latest,${img}:${{ github.sha }}" ;;
```
I wonder if we should also tag release images with the release version as well as the sha and latest?

stevenhorsman commented on .github/workflows/podvm_binaries.yaml:
```diff
@@ -57,12 +57,25 @@ jobs:
+
+        case "$event_name" in
+        release)
+          tags="${img}:latest,${img}:${{ github.sha }}" ;;
```
Same question about release version

stevenhorsman reviewed (commented):
Generally it looks very good (if we agree to delay caching to a later PR, which I do), I had a few comments/questions

wainersm commented:
hi @stevenhorsman ! Thanks for the prompt review!

I addressed your comments:
* renamed podvm_on_release -> podvm_publish as the pipeline can be triggered manually too. In the future it can be used to publish the image on `push`. Tested workflow_dispatch here: https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/6199099196 (ps: 49min to complete the entire build is insane...)
* The images on release will now be tagged with the git tag too (partially fixing https://github.com/confidential-containers/cloud-api-adaptor/issues/1109 I guess).
* On push and workflow_dispatch the image will be tagged with SHA and `latest`

wainersm commented:
ah, regarding caching...I am making some experiments to use the `cache-to` and `cache-from` argument of `docker build`....will be sent in a separate PR.

stevenhorsman reviewed (approved):
LGTM. Thanks!

snir911 commented on .github/workflows/podvm.yaml:
```diff
@@ -55,12 +55,31 @@ jobs:
+        pull_request|pull_request_target)
+          tags="${img}:ci-pr${{ github.event.number }}" ;;
+        push|workflow_dispatch)
+          tags="${img}:latest,${img}:${{ github.sha }}" ;;
```
workflow_dispatch is manual executing no? do we want to push to latest i this case?

wainersm replied:
@snir911 yes, `workflow_dispatch` is manually executing. With current workflows we some times (very rare I'd say) manually build the images from main and they should get labeled `latest`

snir911 commented on .github/workflows/podvm.yaml:
```diff
@@ -100,3 +117,4 @@ jobs:
           "UBUNTU_IMAGE_CHECKSUM="
           "BINARIES_IMG=${{ inputs.registry }}/podvm-binaries-${{ matrix.os }}-${{ matrix.arch }}:${{ inputs.binaries_img_tag }}"
           "BUILDER_IMG=${{ inputs.registry }}/podvm-builder-${{ matrix.os }}:${{ inputs.builder_img_tag }}"
+          "CAA_SRC_REF=${{ env.CAA_SRC_REF }}"
```
note that there's also https://github.com/confidential-containers/cloud-api-adaptor/pull/1448 PR ongoing, IIUC it will allow you simplify this series

wainersm replied:
yep, @jtumber-ibm has also sent another PR to avoid rebuilding the agent all the time. Plenty of work is on the queue to simplify/speed up the build process but this PR is considering the old way.

snir911 reviewed (commented):
Thanks Wainer! I added couple of comments.

ldoktor commented on .github/workflows/podvm_binaries.yaml:
```diff
@@ -57,12 +57,31 @@ jobs:
         username: ${{ github.repository_owner }}
         password: ${{ secrets.GITHUB_TOKEN }}
 
+    - name: Determine image tags
```
Shouldn't the `GITHUB_ENV` be passed across downstream jobs? If I understand it correctly then this step should be only present in `podvm_builder` and the remaining steps should simply rely on it (I'd like to avoid duplicating the embedded steps)

ldoktor replied:
Hmm, I can see later that you perhaps want these to be callable separately. Not sure if it's a benefit as it need quite a lot of code-duplication. Anyway let me know what you think about the `GITHUB_ENV` or `INPUTS` use for these as the steps depend on each-other anyway so I don't think they are likely to be called with different arguments.

wainersm replied:
hi @ldoktor ! First of all, thanks for reviewing this PR.

I did not understand your idea, though.

Sending a variable to `$GITHUB_ENV` ensures the next steps within same job will have the variable exported on the environment.

In this PR, I made the `podvm_builder`, `podvm_binaries` and `podvm` workflows callable so they can be used by different workflows. And the way to passes values to those callable workflows is via `inputs`. If there is a way o inherent environment variables between jobs I am not aware of.

ldoktor replied:
I'm not sure if `GITHUB_ENV` could be passed between jobs so you are most certainly right. But the `INPUTS` should work. Not sure if it would result in less boiler-plate, though.

ldoktor replied:
I mean it's not a deal-breaker, I just dislike the amount of repeated work. If we constructed the `CAA_SRC_REF` in the first workflow and passed only the product, we might remove most of the duplicities.

ldoktor commented on .github/workflows/podvm.yaml:
```diff
@@ -40,6 +44,19 @@ jobs:
+        [ -n "$caa_src_ref" ] || \
+          caa_src_ref="$(yq '.git.cloud-api-adaptor.reference' versions.yaml)"
+        [ -n "$caa_src_ref" ]
+        echo "CAA_SRC_REF=${caa_src_ref}" >> $GITHUB_ENV
```
The same question goes here, isn't it possible to forward the `CAA_SRC_REF` to the downstream steps?

wainersm replied:
`echo "CAA_SRC_REF=${caa_src_ref}" >> $GITHUB_ENV` ensures `CAA_SRC_REF` is forwarded to downstream steps **on same job** . Do you mean between jobs?

ldoktor commented on .github/workflows/podvm.yaml:
```diff
@@ -40,6 +44,19 @@ jobs:
+    - name: Read properties from versions.yaml
+      run: |
+        # There reference to CAA sources will honored if passed to this
+        # workflow via inputs.
```
I'm not native but this sentence is quite hard for me to digest. How about `Allow to override the CAA version from versions.yaml from caa_src_ref input`? (again, not I'm native and not well familiar with CAA)

wainersm replied:
Re-reading what I wrote, I didn't even understood myself :D

Gonna accept your suggestion.

ldoktor commented on podvm/Dockerfile.podvm_builder:
```diff
@@ -63,7 +63,15 @@ RUN echo $CAA_SRC
 
-RUN git clone ${CAA_SRC} -b ${CAA_SRC_REF} cloud-api-adaptor
+RUN if [ -n "${CAA_SRC}" ]; then \
+      rm -rf cloud-api-adaptor && \
```
Hmm, is is likely the `cloud-api-adaptor` is present already? If so dropping everything via `rm -rf` seems like a waste to me as many objects can be reused. How about `git remote set-url origin ${CAA_SRC}`?

wainersm replied:
Reset the remote url (and perhaps fetch the branch again?) seems a good idea. There a couple of places where we delete the repository and clone again, where we can use that idea. However, I would prefer to send in a separate PR as this change will need to re-test this PR (which is not an easy task :( Is it okay for you?

ldoktor replied:
Sure, it's just an optimization

wainersm commented:
I just updated the PR to re-word some comments as suggested by @ldoktor. Everything else is the same.

ldoktor commented:
Seems alright, the git `rm -rf` optimization should be handled separately (if needed). As for the repeated image tags handling I don't have enough experience with GH actions to say for sure so it's probably fine, but if there is a possibility to share the code or to just accept produced variables instead of re-typing the same code again and again, I'd welcome that. But more important is that functional-wise it seems to be working well so unless you find it easy to address the duplication, it's fine.

stevenhorsman commented:
@wainersm, once this is refactored, do you think we need this in to help with the 0.8 release testing? It would be great to merge it sooner rather than later and get more regular CI running as we've established that the podvm builds have gone a bit wrong for centos lately

wainersm commented:
In reason of the recent changes on the podvm building to allow using the local source tree, I had to re-do most of this PR. Actually it dropped some commits and adapted others, in general the implementation is much more simpler now. The logic is the same though: build the podvm images tagged with ci-prNUMBER then push to ghcr.io. 

@ldoktor mind to have another look? Most of the boilerplate are gone
@stevenhorsman If you find (you always find!) some time to review it... :)

I tried to test it but the podvm_binaries build has failed (tried 2x) due to networking issue as you can see in https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/6881815692/job/18720954383?pr=63

I hope later today or tomorrow that networking issue is solved so I can give it a 3rd tentative.

wainersm commented:
> In reason of the recent changes on the podvm building to allow using the local source tree, I had to re-do most of this PR. Actually it dropped some commits and adapted others, in general the implementation is much more simpler now. The logic is the same though: build the podvm images tagged with ci-prNUMBER then push to ghcr.io.
> 
> @ldoktor mind to have another look? Most of the boilerplate are gone @stevenhorsman If you find (you always find!) some time to review it... :)
> 
> I tried to test it but the podvm_binaries build has failed (tried 2x) due to networking issue as you can see in https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/6881815692/job/18720954383?pr=63
> 
> I hope later today or tomorrow that networking issue is solved so I can give it a 3rd tentative.

ok, re-ran the job today, all podvm passed except for CentOS Stream which I already expected to fail. Here is: https://github.com/wainersm/cc-cloud-api-adaptor/actions/runs/6881815692/job/18745812378

stevenhorsman commented on .github/workflows/podvm_publish.yaml:
```diff
@@ -0,0 +1,27 @@
+on:
+  # Create release Pod VM Images
+  release:
+    types: [created]
```
This is fine for now, but I wonder if longer term we want to have a master release workflow that calls this via `workflow_call` so we can do things like trigger the CAA build and release tagging at the same time?

wainersm replied:
hi @stevenhorsman ,

By "release tagging" you meant tagging the image (e.g `quay.io/confidential-containers/cloud-api-adaptor:v0.9.0` and `quay.io/confidential-containers/podvm-generic-ubuntu-amd64:v0.9.0`)?

I was thinking in send a follow up PR to tag the podvm image, yes. I wasn't thinking about the CAA image, it seems a good idea btw

stevenhorsman reviewed (approved):
Looking good. Thanks for the work x2!

ldoktor reviewed (approved):
Yep, the changes make sense to me, naming seems corresponding to `Makefile` and it looks nicer then the previous version.

stevenhorsman commented:
So not something for this PR to resolve, but I've been trying out running the libvirt e2e flow on s390x and hit problems with the podvm container image. It looks like the image we push for s390x isn't properly multi-arch:
```
# docker manifest inspect quay.io/confidential-containers/podvm-generic-ubuntu-s390x:6df29bf2dcb8f332327732fe91e457566c231ba2902f06474aa7f4d001d7b20f
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.index.v1+json",
   "manifests": [
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "size": 482,
         "digest": "sha256:ff0100c21c90897a2deee74840f92f635bfeed8f6159dd36050756106941c5d7",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "size": 566,
         "digest": "sha256:16adb91d0b26e1628c3636fe71faa2334ef333c49cf7220306a85b715544e8f4",
         "platform": {
            "architecture": "unknown",
            "os": "unknown"
         }
      }
   ]
}
root@sh-libvirt-s390x-e2e-test:~/go/src/github.com/confidential-containers/cloud-api-adaptor/podvm# docker pull quay.io/confidential-containers/podvm-generic-ubuntu-s390x:6df29bf2dcb8f332327732fe91e457566c231ba2902f06474aa7f4d001d7b20f
6df29bf2dcb8f332327732fe91e457566c231ba2902f06474aa7f4d001d7b20f: Pulling from confidential-containers/podvm-generic-ubuntu-s390x
no matching manifest for linux/s390x in the manifest list entries
```
so when we do the `docker create --name podvm-exporter-2023111714531700232811 quay.io/confidential-containers/podvm-generic-ubuntu-s390x:6df29bf2dcb8f332327732fe91e457566c231ba2902f06474aa7f4d001d7b20f /bin/sh` step in ./hack/download-image.sh it doesn't work

wainersm commented:
> Yep, the changes make sense to me, naming seems corresponding to `Makefile` and it looks nicer then the previous version.

Thanks @ldoktor !

snir911 reviewed (approved):
Thanks! LGTM
